core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -96,7 +117,15 @@ public long getSizeInBytes()
 
     private void appendDictionary(IntArrayList positions, DictionaryBlock source)
     {
-        delegate.append(mapPositions(positions, source), source.getDictionary());
+        Block dictionary = source.getDictionary();
+        IntArrayList mappedPositions = mapPositions(positions, source);
+        if (dictionaryBlockBuilder.canAppend(dictionary)) {
+            dictionaryBlockBuilder.append(mappedPositions, dictionary);
+        }
+        else {
+            dictionaryBlockBuilder.flushDictionary(delegate);
+            delegate.append(mappedPositions, dictionary);
+        }
```

Comment:
Can the `dictionaryBlockBuilder.append` internally handle the change in dictionary ?

Reply:
we do not want multiple dictionaries but we could move the above `if` to the dictionaryBlockBuilder. IMO it better fits here.